e2e/nomostest/gitproviders/repository.go

```diff
@@ -51,193 +52,223 @@ const (
 		SafetyNSName:          safetyName,
 		SafetyNSPath:          fmt.Sprintf("acme/namespaces/%s/ns.yaml", safetyName),
 		SafetyClusterRoleName: safetyName,
 		SafetyClusterRolePath: fmt.Sprintf("acme/cluster/cluster-role-%s.yaml", safetyName),
-		Scheme:                nt.Scheme,
-		Logger:                nt.Logger,
-		GitProvider:           nt.GitProvider,
+		Scheme:                scheme,
+		Logger:                logger,
+		GitProvider:           provider,
+		PrivateKeyPath:        privateKeyPath,
+		yamlSerializer:        jserializer.NewYAMLSerializer(jserializer.DefaultMetaFactory, scheme, scheme),
 	}
+}
 
-	repoName, err := nt.GitProvider.CreateRepository(namespacedName)
-	// Add the repo to nt.RemoteRepositories immediately after it is created to reuse the repo.
-	nt.RemoteRepositories[nn] = g
+// Create the remote repository using the GitProvider, and
+// create the local repository with an initial commit.
+func (g *Repository) Create() error {
+	repoName, err := g.GitProvider.CreateRepository(g.Name)
 	if err != nil {
-		nt.T.Fatal(err)
+		return errors.Wrapf(err, "creating repo: %s", g.Name)
 	}
 	g.RemoteRepoName = repoName
 
-	g.init(nt.gitPrivateKeyPath)
-	g.initialCommit(sourceFormat)
+	// TODO: Move this to the caller or GitProvider
+	// Add the repo to nt.RemoteRepositories immediately after it is created to reuse the repo.
+	// nt.RemoteRepositories[nn] = g
 
-	return g
+	return g.Reformat(g.Format)
 }
 
-// ReInit re-initializes the repo to the initial state.
-func (g *Repository) ReInit(nt *NT, sourceFormat filesystem.SourceFormat) {
-	nt.T.Helper()
-
-	// Update test environment
-	g.T = nt.T
-	// Reset repo contents
-	g.init(nt.gitPrivateKeyPath)
-	g.initialCommit(sourceFormat)
+// Reformat the repo using the specified source format.
+// Use Repository.Format to keep the format the same.
+func (g *Repository) Reformat(format filesystem.SourceFormat) error {
+	if err := g.init(); err != nil {
+		return err
+	}
+	return g.initialCommit(format)
 }
 
-func (g *Repository) gitCmd(command ...string) *exec.Cmd {
+// Git executes the command from the repo root.
+// The command is always logged to the debug log.
+// If the command errors, the command and output is logged.
+func (g *Repository) Git(command ...string) ([]byte, error) {
 	// The -C flag executes git from repository root.
 	// https://git-scm.com/docs/git#Documentation/git.txt--Cltpathgt
 	args := []string{"git", "-C", g.Root}
 	args = append(args, command...)
 	g.Logger.Debugf("[repo %s] %s", path.Base(g.Root), strings.Join(args, " "))
-	return exec.Command(args[0], args[1:]...)
-}
-
-// Git wraps shelling out to git, ensuring we're running from the git repository
-//
-// Fails immediately if any git command fails.
-func (g *Repository) Git(command ...string) {
-	g.T.Helper()
-
-	cmd := g.gitCmd(command...)
-	out, err := cmd.CombinedOutput()
+	out, err := exec.Command(args[0], args[1:]...).CombinedOutput()
 	if err != nil {
+		if !g.Logger.IsDebugEnabled() {
```

Comment:
If debug is not enabled - is this right? I'd think we'd want to log the extra information if debug is enabled, or just `Logger.Debugf`

Reply:
The command is printed to the debug log BEFORE the call. So if debug is NOT enabled, we need to print the command AFTER the call errors.

This is how it works elsewhere for command failure. Alternatively we could just always print the command to the normal log before the call.

Reply:
Ah I missed that, sounds good then